wxDVC: support DnD for multiple formats at once.

laptabrok opened the pull request:
In generic wxDVC `EnableDropTarget( const wxDataFormat &format )` method worked correctly only with formats which internal data may be copied with `memcpy` (`wxCustomDataObject `restriction). Moreover it enabled support for only one format at the same time, althouh method's name means to _enable_ support for format w/o removing support for other _enabled_ formats (btw, maybe I'm wrong).

There is support for multiple formats at once with this patch. (What about `disabling `specific format?).

Also there are at least two unused vars in code: `m_dropEnabled`, `m_dropFormat`, do we really need them?

When we are ready for the wxEVT_DATAVIEW_ITEM_DROP event, I place dropObject (`wxDataObjectComposite`) into event and still copy the data to event, do we really need it? We always can copy data from dropObject if needed. The same happens in the wxOSX port.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -2084,11 +2084,61 @@ bool wxDataViewMainWindow::EnableDragSource( const wxDataFormat &format )
+    wxDataObjectComposite *dataObject;
+    wxDataViewDropTarget *dropTarget = static_cast<wxDataViewDropTarget*>(GetDropTarget());
+    if (dropTarget)
+        dataObject = static_cast<wxDataObjectComposite*>(dropTarget->GetDataObject());
```
```suggestion
    {
        dataObject = static_cast<wxDataObjectComposite*>(dropTarget->GetDataObject());
    }
```
Just for consistency with the `else` branch,

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -2084,11 +2084,61 @@ bool wxDataViewMainWindow::EnableDragSource( const wxDataFormat &format )
+        return false;
+
+    wxDataObjectComposite *dataObject;
+    wxDataViewDropTarget *dropTarget = static_cast<wxDataViewDropTarget*>(GetDropTarget());
```
```suggestion
    wxDropTarget* const dropTarget = GetDropTarget();
```
It doesn't look like we actually need the cast to the derived class, do we?

laptabrok replied:
Yes, we can do w/o casting.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -2176,12 +2226,18 @@ wxDragResult wxDataViewMainWindow::OnData( wxDataFormat format, wxCoord x, wxCoo
-    event.SetDataBuffer( obj->GetData() );
+    event.SetDataSize(dataSize);
+    event.SetDataBuffer(buffer.GetData());
+    event.SetDataObject(obj);
```
I wonder if it could be better to pass `obj->GetObject(format)` here.

laptabrok replied:
In wxOSX port `wxCompositeObject` is passed into event, so I decided made it the same way in Generic for compatibility, but agree, it is better to pass `obj->GetObject(format)`.

vadz commented on src/generic/datavgen.cpp:
```diff
@@ -2176,12 +2226,18 @@ wxDragResult wxDataViewMainWindow::OnData( wxDataFormat format, wxCoord x, wxCoo
-    wxCustomDataObject *obj = (wxCustomDataObject *) GetDropTarget()->GetDataObject();
+    wxDataObjectComposite *obj = static_cast<wxDataObjectComposite*>(GetDropTarget()->GetDataObject());
+
+    size_t dataSize = obj->GetDataSize(format);
+    wxMemoryBuffer buffer;
+    obj->GetDataHere(format, buffer.GetWriteBuf(dataSize));
+    buffer.UngetWriteBuf(dataSize);
```
Data could be relatively big here, I wonder if it could be worth it to retrieve it only on demand from `wxDataViewEvent::GetDataBuffer()` instead of doing it like this.

laptabrok replied:
Ok, I can update `wxDataViewEvent::GetDataBuffer()` method.

vadz reviewed (commented):
AFAICS `EnableDropTarget()` only handles a single format because calling `gtk_tree_view_enable_model_drag_dest()` again seems to replace/override the previously set targets. Moreover, it's going to be a bit tricky (although possible) to add support for multiple formats to it later if we do it like this PR does it, i.e. allowing to call `EnableDropTarget()` multiple times.

IMO it would be simpler to just have `EnableDropTarget()` overload taking a vector of formats (the existing version would then simply forward to it with a single-element vector or call some new `DisableDropTarget()` if the format is invalid). Would this be suitable for your needs? If so, I think it would be worth redoing the PR to do it like this because it would also address your (perfectly legitimate question) about disabling some formats.

I also agree that we can get rid of `m_dropEnabled` and `m_dropFormat`, but this can/should be done in a separate commits.

Also, could you please update the documentation in `interface/wx/dataview.h` to describe the new behaviour? New functions should have `@since 3.1.4` tag in them and documentation of MSW-specific functionality should mention that it's not available in the native GTK/Mac versions.

TIA!

laptabrok committed:
wxDVC improved to accept of multiple data formats dragged on the same window.

wxDataViewCtrlBase::EnableDropTarget is not virtual now and overloaded to accept array of wxDataFormat to to handle different formats dropped on the same DVC.
Protected virtual method 'DoEnableDropTarget` presented to be overrided in each WX port. The default implementation does nothing.
Passing empty array or wxDF_INVALID format disables DnD operations at all for selected DVC.

Native macOS implementation refactored: all operations with dragged data are handled by wxDropTarget. Private unnecessary 'DataViewPboardType' data type removed.

Sample 'dataview' updated for new interface.

laptabrok commented:
I'm working on wxDVC's DnD in wxOSX port now. By default it can accept wxDF_TEXT , wxDF_UNICODETEXT. I added support for wxDF_FILENAME. On wxOSX DVC DnD enabled by default w/o EnableDropTarget call. Don't know about other formats (wxDF_BITMAP, wxDF_HTML etc.) but I can try to implement EnableDropTarget method for explicitly enabling DnD for wxDVC on wxOSX port. Also I can update packing of event structure here (do not pass buffer data, pass simpleDataObject instead of dataObjectComposite etc.) as we discussed here.

vadz commented:
I agree that it would be better if Mac version required calling `EnableDropTarget()`, just as the other ones do, even if it's formally a compatibility breakage. The other changes are welcome too, although it would be best to make them in separate commits. TIA!

Also please note the wxQt build error, which actually shows a real problem: you need to use `GetType()`, current code doesn't work correctly for wxGTK neither.

laptabrok commented:
Implemented EnableDropTarget with array of wxDataFormat.
Implemented EnableDropTarget for wxOSX (supports text and filenames).
Typed wxDataObjectSimple passed to event instead of wxDataObjectComposite on Generic and wxOSX.
Data from object is not copied to event.
Fixed QT build with `GetType()`.

laptabrok commented on include/wx/osx/cocoa/dataview.h:
```diff
@@ -549,6 +549,8 @@ class wxCocoaDataViewControl : public wxWidgetCocoaImpl,
 
     virtual void SetFont(const wxFont& font, const wxColour& foreground, long windowStyle, bool ignoreBlack = true);
 
+    const wxCocoaOutlineView* GetView() { return m_OutlineView; }
```
Is it a good solution? We need view to get its registered dragged types.

laptabrok commented on src/osx/cocoa/dataview.mm:
```diff
@@ -819,35 +820,38 @@ -(NSDragOperation) setupAndCallDataViewEvents:(wxEventType)eventType dropInfo:(i
-        ::CFRelease(osxData);
-        delete dataObjects;
+        wxOSXPasteboard wxPastboard(pasteboard);
+        if (dataObject->ReadFromSource(&wxPastboard))
```
This approach seems to be invalid in case of dragging text. 😔 I received only the first letter of dragged Latin string or invalid symbols of dragged Cyrillic string.

vadz replied:
Looks like a wrong cast from `wchar_t*` to `char*` somewhere.

laptabrok replied:
@csomor can you help me with this issue, please? Can I use `ReadFromSource` in case of dragging text here?

csomor replied:
I'll try to look at it over the weekend

csomor replied:
usually when using 'internal sources' I don't go via the Pasteboard at all, I call the `wxDropSource::GetCurrentDropSource` and then read from that object (see `dnd_osx.cpp`, `wxDropTarget::GetData()`), I'll dive deeper

(btw. please keep the structure of having an else branch and no early return in the if branch)

csomor replied:
ok I see that dataview.mm is setting up the strings itself (I didn't write dataview.mm), what are the steps I need to perform to view the problem exactly ? Thanks

csomor replied:
I found the bug, the code is assuming we are always using UTF8 internally, in line 927 in dataview.mm:

```
 itemString = wxString::FromUTF8(static_cast<char const*>(dataBuffer.GetData())+sizeof(wxDataFormatId),dataSize)
```

csomor replied:
I'd suggest replacing this line with something along, @vadz is this fine for you ?
```
static wxMBConvUTF16 s_UTF16Converter;
itemString = wxString( static_cast<char const*>(dataBuffer.GetData())+sizeof(wxDataFormatId), s_UTF16Converter, dataSize );
```

laptabrok replied:
> ok I see that dataview.mm is setting up the strings itself (I didn't write dataview.mm), what are the steps I need to perform to see the problem exactly ? Thanks

You can run the dataview sample and do dnd inside the app and drop text from outside.

csomor replied:
Thanks, I found out I could drag Beethoven around :-)

vadz replied:
> 
> 
> I'd suggest replacing this line with something along this, @vadz is this fine for you ?
> 
> ```
> static wxMBConvUTF16 s_UTF16Converter;
> itemString = wxString( static_cast<char const*>(dataBuffer.GetData())+sizeof(wxDataFormatId), s_UTF16Converter, dataSize );
> ```

Sorry, I didn't really have time to look at this code yet, but why do we need a static converter? It shouldn't have any state. More generally, isn't there already some function in wxMac to do this conversion?

csomor replied:
> Sorry, I didn't really have time to look at this code yet, but why do we need a static converter? It shouldn't have any state. More generally, isn't there already some function in wxMac to do this conversion?

No hard reason, only the allocation, as we do so with wxMBConvUTF16 at other places in wx, I had just copied this. 

I have not found another place where I'm doing a UTF16 to wxString in osx code.

laptabrok replied:
> I'd suggest replacing this line with something along this, @vadz is this fine for you ?
> 
> ```
> static wxMBConvUTF16 s_UTF16Converter;
> itemString = wxString( static_cast<char const*>(dataBuffer.GetData())+sizeof(wxDataFormatId), s_UTF16Converter, dataSize );
> ```

Everything works!

vadz commented:
Hmm, looks like you've unchecked the "Allow edit from maintainers" for this PR as I get
```
ERROR: Permission to eligovision/wxWidgets.git denied to vadz.
```
when trying to push to this branch, so I had to create #1912 to test the build after merging it with master (there were a couple of conflicts, I hope I've resolved them correctly, but please check).

If you'd like to, you could fast-forward your `upstream-master` branch (which is a really confusing name for a branch which is not master and doesn't belong to the upstream...) to my `dvc-dnd` branch. Otherwise, please close this one and let's switch to #1912.

laptabrok commented:
@vadz unfortunately i could not push to your branch getting the same error as you got trying push to mine.

vadz commented:
> @vadz unfortunately i could not push to your branch getting the same error as you got trying push to mine.

This is normal, as you're not a maintainer of this project. But it seems that me being unable to push into your fork is normal too because it belongs to an organization and isn't just your personal fork.

Anyhow, thanks for the update, I'll try to have a look at the latest changes a.s.a.p. (assuming the CI builds pass).

laptabrok commented on src/osx/cocoa/dataview.mm:
```diff
@@ -820,36 +813,34 @@ -(NSDragOperation) setupAndCallDataViewEvents:(wxEventType)eventType dropInfo:(i
-        else
-            delete textDataObject;
-        // send event if data could be copied:
+        if ([bestType compare:NSFilenamesPboardType] == NSOrderedSame)
+        {
+            dataObject = new wxFileDataObject;
+        }
+        else if ([bestType compare:NSStringPboardType] == NSOrderedSame)
+        {
+            dataObject = new wxTextDataObject;
+        }
+        // TODO check for other formats here?
```
I covered only `NSFilenamesPboardType` and `NSStringPboardType` here. Not sure if the rest of the types specified in the `EnableDropTarget` will work.

csomor replied:
if you move the 'generic' EnableDropTarget as suggested below, then you can make the else branch start like this

```
    else
    {
        dataObjects = new wxDataObjectComposite();
        wxDataFormat wxFormat((wxDataFormat::NativeFormat)bestType);
        switch (wxFormat.GetType() )
        {
            case wxDF_FILENAME :
                dataObject = new wxFileDataObject;
                break;
            case wxDF_TEXT :
            case wxDF_UNICODETEXT:
                dataObject = new wxFileDataObject;
                break;
            default:
                break;
        }

```

That way all modern UTI types (the PboardTypes you were using will be deprecated) are supported automatically, and you could move this code again to a superclass ...

laptabrok commented on src/osx/cocoa/dataview.mm:
```diff
@@ -2552,6 +2545,72 @@ void UpdateWithRow(int row)
+            }
+        }
+
+        // Add support for internal dataView items' DnD?
+        // CFArrayAppendValue(typesarray, DataViewPboardType);
```
Should we enable `DataViewPboardType`? Maybe when enabling support for text types? `DataViewPboardType` is used in `writeItems` (line 849) when item's drag begin (BTW, I quite not understand the meaning of its use. Maybe Is it used for dragging items outside the app?

csomor replied:
`DataViewPboardType` is the private data type (=string constant) of the dataview implementation. It has the wxDataFormat ID in its first bytes, then the raw data

csomor replied:
> `DataViewPboardType` is the private data type (=string constant) of the dataview implementation. It has the wxDataFormat ID in its first bytes, then the raw data, only used within the app / wx apps themselves

laptabrok commented on src/osx/cocoa/dataview.mm:
```diff
@@ -2552,6 +2545,72 @@ void UpdateWithRow(int row)
+    {
+        wxCFMutableArrayRef<CFStringRef> typesarray;
+        for (size_t i = 0; i < formats.GetCount(); ++i)
+            // formats.Item(i).AddSupportedTypes(typesarray);
```
AFAIU, this adds osx carbon types instead of modern cocoa ones.

csomor replied:
NSStringPboardType is an alias for a NSString which is toll free bridged to a CFStringRef, the wxCFMutableArray can be cast to an NSArray* so this is fine

laptabrok replied:
@csomor, I mean that with just calling `formats.Item(i).AddSupportedTypes(typesarray);` in for-loop (w/o my swicth) DnD does not work at all. Am i doing something wrong?

laptabrok commented on src/gtk/dataview.cpp:
```diff
@@ -3720,9 +3720,16 @@ bool wxDataViewCtrlInternal::EnableDragSource( const wxDataFormat &format )
+bool wxDataViewCtrlInternal::EnableDropTarget( const wxDataFormatArray& formats )
 {
-    wxGtkString atom_str( gdk_atom_name( format  ) );
+    if (formats.GetCount() == 0)
+    {
+        // TODO remove drop target if exists
+
+        return true;
+    }
```
Need help with implementation here.

laptabrok commented:
@vadz, I also noticed that during drag processing we don't know anything about dragging item, if we dragging one. `DropItemInfo` contains info about target item. i.e. destination. Will be a good approach to store dragging item in `DropItemInfo` and also pass it to `wxDataViewEvent` to be able do additional checks when processing `wxEVT_DATAVIEW_ITEM_DROP_POSSIBLE` event and others? It won't take much memory, but also will give us to understand that we are within the same widget

csomor commented:
@laptabrok right now you have a 'generic' implementation in `wxDataViewMainWindow::EnableDropTarget( const wxDataFormatArray &formats )` Could you move this to base, so that it can be used as a default throughout ? a special implementation on OSX wouldn't be necessary anymore, `SetDropTarget` on osx already is capable of handling a composite data object, that way you'd also get all conversions that are built-in

csomor commented:
> @laptabrok right now you have a 'generic' implementation in `wxDataViewMainWindow::EnableDropTarget( const wxDataFormatArray &formats )` Could you move this to base, so that it can be used as a default throughout ? a special implementation on OSX wouldn't be necessary anymore, `SetDropTarget` on osx already is capable of handling a composite data object, that way you'd also get all conversions that are built-in, eg for text and file:

Sorry, I was overlooking the built-in private type, which does not get routed via a wx API, so this method should be something along

```
bool wxCocoaDataViewControl::EnableDropTarget(const wxDataFormatArray& formats)
{
    [m_OutlineView unregisterDraggedTypes];

    if (formats.GetCount() > 0)
    {
        wxCFMutableArrayRef<CFStringRef> typesarray;
        for (size_t i = 0; i < formats.GetCount(); ++i)
            formats.Item(i).AddSupportedTypes(typesarray);

        // Add support for internal dataView items' DnD
        CFArrayAppendValue(typesarray, DataViewPboardType);

        [m_OutlineView registerForDraggedTypes:typesarray];
    }

    return true;
}

```

vadz commented:
@csomor Stefan, do you have any comments about this? I am not sure if this is really 3.1.4-critical, but if you're ok with the changes here on the Mac side, I could try to integrate it before the release, please let me know.

csomor commented:
part of the mac changes are not really related to the PR, but rather to bring things in line with the dealing of dataobj in the rest of wx, I could do this as well, but I don't know how to make this properly without ruining this PR's build

vadz commented:
@csomor Yes, this is a bit confusing as there are many changes here, but looking at this again, it doesn't seems to be finished yet (e.g. there are TODO comments in the code). I think you should try to extract whichever parts help wxOSX and are not related to multiple formats support and then the author should rebase the PR on the latest master because we're clearly not going to merge all the existing 27 commits as they're anyhow, i.e. it's not a problem to lose the current commit history (and if you (@laptabrok) have any problems with that, I could do it instead, but this would require opening a new PR because I can't push to your org fork of wx).

csomor commented:
ok, I'll do that, thanks

laptabrok commented:
> @csomor Yes, this is a bit confusing as there are many changes here, but looking at this again, it doesn't seems to be finished yet (e.g. there are TODO comments in the code). I think you should try to extract whichever parts help wxOSX and are not related to multiple formats support and then the author should rebase the PR on the latest master because we're clearly not going to merge all the existing 27 commits as they're anyhow, i.e. it's not a problem to lose the current commit history (and if you (@laptabrok) have any problems with that, I could do it instead, but this would require opening a new PR because I can't push to your org fork of wx).

Sorry for long delay, I'll do it tomorrow.

laptabrok commented:
> > @csomor Yes, this is a bit confusing as there are many changes here, but looking at this again, it doesn't seems to be finished yet (e.g. there are TODO comments in the code). I think you should try to extract whichever parts help wxOSX and are not related to multiple formats support and then the author should rebase the PR on the latest master because we're clearly not going to merge all the existing 27 commits as they're anyhow, i.e. it's not a problem to lose the current commit history (and if you (@laptabrok) have any problems with that, I could do it instead, but this would require opening a new PR because I can't push to your org fork of wx).
> 
> Sorry for long delay, I'll do it tomorrow.

Should I just rebase all my changes or I need to split commits by platforms (btw I marked corresponding port in each commits). Also there are no commits that are not related dnd improvements: I had to change some classes' interfaces, but all was done beacause of DnD only. Also TODOs are related for some platform-specific parts and do not breake old behavior.

vadz commented:
Sorry for the delay with replying. It's not necessary to have separate commits for each platform, although it could be done like this if you prefer. It would be nice to squash all the "Update xxx" and other "fixup"-like commits together, however, could you please do it? TIA!

laptabrok committed:
Doc for wxDVC interfaces updated.

vadz commented:
Thanks for the updates! Is this PR considered to be complete/ready for the review now?

laptabrok commented:
> Thanks for the updates! Is this PR considered to be complete/ready for the review now?

Yes, I squashed and rebased it for convinience and updated the Doc.

I moved common code for data object creation to separate protected virtual method `wxDataViewCtrlBase::CreateDataObject` to be used in any port. By default it constructs `wxDataObjectComposite` from passed `wxDataFormatArray` if it has at least one data format, otherwise it returns `NULL`.

It would be great to move `wxDataViewCtrl::DoEnableDropTarget` from macOS port (src/osx/dataview_osx.cpp) to `wxDataViewCtrlBase` which can be default behaviour, and override it in generic version (it has its own private `wxDataViewDropTarget`), but it won't compile because `wxDropTarget::OnData` is pure virtual in generic (I noticed that in any other wx port `wxDropTarget` has all virtual methods implemented so it may be a good reason for new PR and update wxDVC ;) ).

In native GTK implementation the only one format at once supported yet although I found in [doc](https://developer.gnome.org/gtk3/stable/GtkTreeView.html#gtk-tree-view-enable-model-drag-dest) that GTK3 supports DnD for multiple formats via passing C-style array of `GtkTargetEntry`. But I work only on Windows and macOS, so someone else can implement it...

Stefan @csomor, I rewritten macOS part to handle data via its native `wxDropTarget` instead manually way. `setupAndCallDataViewEvents` does all the work when somethig dragged/dropped. `writeItems` also has new simple design. Methods `callDataViewEvents`, `wxCocoaDataViewControl::GetDnDDataFormat` and `wxCocoaDataViewControl::GetDnDDataObjects` have been removed.

laptabrok commented on src/gtk/dataview.cpp:
```diff
@@ -3814,9 +3814,16 @@ bool wxDataViewCtrlInternal::EnableDragSource( const wxDataFormat &format )
+    wxGtkString atom_str( gdk_atom_name( formats.Item(0) ) );
     m_dropTargetTargetEntryTarget = wxCharBuffer( atom_str );
 
     m_dropTargetTargetEntry.target =  m_dropTargetTargetEntryTarget.data();
```
Here we can create C-style `GtkTargetEntry` array instead of single `m_dropTargetTargetEntry` object and pass it to `gtk_tree_view_enable_model_drag_dest` call.

vadz commented:
I've finally tried returning to this and so rebased the changes here on the latest master, see #2467 so I'm closing this one, but please have a look at my comments there as I can't apply this in the current state. TIA!
